src/read_data_phantom_hdf5.f90

```diff
@@ -0,0 +1,690 @@
+  call read_from_hdf5(massoftypei(1:ntypes),'massoftype',group_id,got,error)
+  call close_hdf5group(group_id,error)
+
+  ndusttypes = ndustsmall + ndustlarge
+
+  !-- Get number of columns in file
+  call get_ncolumns(file_id,ncolstep,ndusttypes,ndustsmall,nptmass)
+
+  !-- Add calculated quantities to data columns as well as 1 for density and 1 for mass
+  ncolumns   = ncolstep + 2 + ncalc
+
+  !--(re)allocate memory
+  if (.not.allocated(dat) .or. npart+nptmass>maxpart .or. ncolumns>maxcol) then
+    call alloc(npart+nptmass,1,ncolumns,mixedtypes=.true.)
+  endif
+
+  !-- Assign things now that memory has been allocated
+  gamma(ifile)                = gammai
+  time(ifile)                 = timei
+  npartoftype(:,ifile)        = 0      ! initialise to zero
+  masstype(:,ifile)           = 0.     ! I don't think this array is actually used anywhere in splash,
+                                       ! since sinks all have different masses anyway
+
+  npartoftype(1,ifile) = npartoftypei(1)                        ! gas
+  npartoftype(2,ifile) = npartoftypei(2)                        ! dust (old)
+  npartoftype(3,ifile) = nptmass                                ! sink
+  npartoftype(4,ifile) = npartoftypei(3)                        ! boundary
+  npartoftype(5,ifile) = npartoftypei(4)                        ! star
+  npartoftype(6,ifile) = npartoftypei(5)                        ! dark matter
+  npartoftype(7,ifile) = npartoftypei(6)                        ! bulge
+  npartoftype(8,ifile) = sum(npartoftypei(7:7+ndusttypes-1))    ! dust
+  npartoftype(9,ifile) = sum(npartoftypei(7+ndusttypes:ntypes)) ! unknown/dead?
+
+  ! npart = sum(npartoftype(1:9,ifile))
+
+  masstype(1,ifile) = massoftypei(1)                        ! gas
+  masstype(2,ifile) = massoftypei(2)                        ! dust (old)
+  ! masstype(3,ifile) = 0.                                    ! sink ???
+  masstype(4,ifile) = massoftypei(3)                        ! boundary
+  masstype(5,ifile) = massoftypei(4)                        ! star
+  masstype(6,ifile) = massoftypei(5)                        ! dark matter
+  masstype(7,ifile) = massoftypei(6)                        ! bulge
+  ! masstype(8,ifile) = 0.                                    ! dust ???
+  ! masstype(9,ifile) = 0.                                    ! unknown/dead ???
+
+  print "(a,i10,a,es10.3,a,i2)",' npart = ',npart,' time = ',time(ifile),' ncolread = ',ncolstep
+
+  !-- Try to read every possible particle array
+  call open_hdf5group(file_id,'particles',group_id,error)
+
+  allocate(itype(npart))
+  call read_from_hdf5(itype,'itype',group_id,got,error)
+  if (got) then
+    iamtype(:,ifile) = 0
+    iamtype(1:npart,ifile) = itypemap_phantom(itype(:))  ! Particles (remap phantom integer labels to splash labels)
+    iamtype(npart+1:npart+nptmass,ifile) = 3             ! Sinks
+  endif
+  deallocate(itype)
+
+  icol = 0
+
+  allocate(array_1d(npart),array_2d(3,npart))
+
```

Comment:
the code from here needs a rewrite. None of the readers in splash explicitly 'know' what they are trying to read from the file, they just read any available datasets of the correct type. All of the hdf5 readers for other formats just query the datasets to find out the number of arrays etc, they do not use the tags to try to match known datasets. This needs to be rewritten, it is unmaintainable as currently written.